OrbitVulkanLayer/DispatchTable.h

```diff
@@ -0,0 +1,413 @@
+      const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdBeginDebugUtilsLabelEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdBeginDebugUtilsLabelEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdEndDebugUtilsLabelEXT CmdEndDebugUtilsLabelEXT(
+      const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdEndDebugUtilsLabelEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdEndDebugUtilsLabelEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdDebugMarkerBeginEXT CmdDebugMarkerBeginEXT(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdDebugMarkerBeginEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdDebugMarkerBeginEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdDebugMarkerEndEXT CmdDebugMarkerEndEXT(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdDebugMarkerEndEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdDebugMarkerEndEXT;
+  }
+
+  template <typename DispatchableType>
+  bool IsDebugMarkerExtensionSupported(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_supports_debug_marker_extension_.contains(key));
+    return device_supports_debug_marker_extension_.at(key);
+  }
+
+  template <typename DispatchableType>
+  bool IsDebugUtilsExtensionSupported(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_supports_debug_utils_extension_.contains(key));
+    return device_supports_debug_utils_extension_.at(key);
+  }
+
+ private:
+  // In vulkan, every "dispatchable type" has as a very first field in memory a pointer to the
+  // internal dispatch table. This pointer is unique per device/instance. So for example
+  // for a command buffer allocated on a certain device, this pointer is the same for the buffer
+  // and for the device. So we can use that pointer to uniquely map dispatchable types to their
+  // dispatch table.
+  template <typename DispatchableType>
+  void* GetDispatchTableKey(DispatchableType dispatchable_object) {
```

Comment:
`dispatchable_object` is passed by value here but by reference in many other places. Is there a reason for that?

Reply:
Well as they are all essentially Vulkan types, and we decided not to pass them as reference, we should not pass them as reference here.
Thanks for spotting the const-refs in all the accessor functions. I removed them and made them pass-by-value.